File: src/main/java/vip/mystery0/base/springboot/service/RedisService.java
package vip.mystery0.base.springboot.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Service;
import vip.mystery0.base.springboot.config.BasePropertiesConfig;
import vip.mystery0.tools.java.factory.JsonFactory;

import java.util.concurrent.TimeUnit;

@Service
@Component
public class RedisService {
    private static final Logger log = LoggerFactory.getLogger(RedisService.class);

    @Autowired
    private BasePropertiesConfig propertiesConfig;
    @Autowired
    private StringRedisTemplate redisTemplate;

    public String getRedisKey(String key) {
        return propertiesConfig.getRedisPrefix() + ":" + key;
    }

    /**
     * 将数据放到redis中
     *
     * @param key   键
     * @param value 值
     */
    public void putIntoRedis(String key, Object value) {
        try {
            String redisKey = getRedisKey(key);
            if (value == null) {
                log.warn("put into redis but value is null, key: {}", redisKey);
                return;
            }
            String redisValue = JsonFactory.toJson(value);
            log.info("put into redis, key: {}, value: {}", redisKey, redisValue);
            redisTemplate.opsForValue().set(redisKey, redisValue);
        } catch (Exception e) {
            log.error("put into redis failed", e);
        }
    }

    /**
     * 从redis中获取值
     *
     * @param key    键
     * @param tClass 返回数据类型
     * @param <T>    数据类型
     * @return 原始数据
     */
    public <T> T getFromRedis(String key,
                              Class<T> tClass) {
        try {
            String redisKey = getRedisKey(key);
            log.info("get key value from redis, key: {}", redisKey);
            String redisValue = redisTemplate.opsForValue().get(redisKey);
            log.info("get value: {}", redisValue);
            if (redisValue == null) {
                return null;
            }
            return JsonFactory.fromJson(redisValue, tClass);
        } catch (Exception e) {
            log.error("get key value from redis error", e);
        }
        return null;
    }

    /**
     * 设置redis键过期时间
     *
     * @param key        键名
     * @param expireTime 过期时间：单位毫秒
     */
    public void updateRedisKeyExpire(String key,
                                     long expireTime) {
        String redisKey = getRedisKey(key);
        log.info("update redis key expire time, key: {}, expire time: {}", redisKey, expireTime);
        try {
            redisTemplate.expire(redisKey, expireTime, TimeUnit.MILLISECONDS);
        } catch (Exception e) {
            log.error("update redis key expire time error, key: {}, expire time: {}", redisKey, expireTime);
        }
    }
}

File: src/main/java/vip/mystery0/base/springboot/service/RedisService.kt
package vip.mystery0.base.springboot.service

import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import vip.mystery0.base.springboot.config.BaseProperties
import vip.mystery0.base.springboot.service.redis.IRedis
import vip.mystery0.base.springboot.utils.withError
import vip.mystery0.tools.kotlin.factory.JsonFactory
import vip.mystery0.tools.kotlin.factory.fromJson
import vip.mystery0.tools.kotlin.factory.toJson
import java.lang.reflect.Type
import java.util.concurrent.TimeUnit
import kotlin.reflect.KClass

/**
 * @author mystery0
 */
@Suppress("unchecked_cast")
@Service
class RedisService(
    objectMapper: ObjectMapper,
    private val properties: BaseProperties,
    private val iRedis: IRedis
) {
    companion object {
        private val log = LoggerFactory.getLogger(RedisService::class.java)
    }

    init {
        JsonFactory.setObjectMapper(objectMapper)
    }

    fun getRedisKey(key: String): String = properties.redisPrefix + ":" + key

    /**
     * 将数据放到redis中
     *
     * @param key   键
     * @param value 值
     */
    fun putIntoRedis(key: String, value: Any?, expireTime: Long = -1L) {
        log.withError("put into redis failed") {
            val redisKey = getRedisKey(key)
            if (value == null) {
                log.warn("put into redis but value is null, key: {}", redisKey)
                return@withError
            }
            val redisValue = value.toJson()
            log.debug("put into redis, key: {}, value: {}, expireTime: {}ms", redisKey, redisValue, expireTime)
            if (expireTime == -1L)
                iRedis.set(redisKey, redisValue)
            else
                iRedis.set(redisKey, redisValue, expireTime, TimeUnit.MILLISECONDS)
        }
    }

    /**
     * 从redis中获取值
     *
     * @param key    键
     * @param type   返回数据类型
     * @param <T>    数据类型
     * @return 原始数据
     */
    fun <T : Any> getFromRedis(key: String, clazz: Class<T>): T? = log.withError<T>("get key value from redis error") {
        val redisKey = getRedisKey(key)
        log.debug("get key value from redis, key: {}", redisKey)
        val redisValue = iRedis.get(redisKey) ?: return@withError null
        log.debug("get value: {}", redisValue)
        if (clazz == String::class.java) return@withError redisValue as T?
        return@withError redisValue.fromJson(clazz)
    }

    /**
     * 从redis中获取值
     *
     * @param key    键
     * @param type   返回数据类型
     * @param <T>    数据类型
     * @return 原始数据
     */
    fun <T : Any> getFromRedis(key: String, clazz: KClass<T>): T? = log.withError<T>("get key value from redis error") {
        val redisKey = getRedisKey(key)
        log.debug("get key value from redis, key: {}", redisKey)
        val redisValue = iRedis.get(redisKey) ?: return@withError null
        log.debug("get value: {}", redisValue)
        if (clazz == String::class) return@withError redisValue as T?
        return@withError redisValue.fromJson(clazz)
    }

    /**
     * 从redis中获取值
     *
     * @param key    键
     * @param type   返回数据类型
     * @param <T>    数据类型
     * @return 原始数据
     */
    fun <T> getFromRedis(key: String, type: Type): T? = log.withError<T>("get key value from redis error") {
        val redisKey = getRedisKey(key)
        log.debug("get key value from redis, key: {}", redisKey)
        val redisValue = iRedis.get(redisKey) ?: return@withError null
        log.debug("get value: {}", redisValue)
        if (type.typeName == String::class.java.typeName) return@withError redisValue as T?
        return@withError redisValue.fromJson(type)
    }

    /**
     * 设置redis键过期时间
     *
     * @param key        键名
     * @param expireTime 过期时间：单位毫秒
     */
    fun updateRedisKeyExpire(key: String, expireTime: Long) {
        val redisKey = getRedisKey(key)
        log.debug("update redis key expire time, key: {}, expire time: {}", redisKey, expireTime)
        log.withError("update redis key expire time error, key: $redisKey, expire time: ${expireTime}ms") {
            iRedis.expire(redisKey, expireTime, TimeUnit.MILLISECONDS)
        }
    }

    /**
     * 删除指定的key
     */
    fun deleteRedisKey(key: String) {
        val redisKey = getRedisKey(key)
        log.debug("delete redis key, key: {}", redisKey)
        log.withError("delete redis key error") {
            iRedis.del(redisKey)
        }
    }
}